### ISSUE TITLE
Allow asynchronous provider ID propagation on NodeClaims

### ISSUE THREAD

**Issue by elmiko** [kubernetes-sigs/karpenter#1273]
### Description

**What problem are you trying to solve?**

On some providers, notably Cluster API, requesting a new instance is an asynchronous process which does not immediately return a provider ID.  The [code for launching nodes in karpenter](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/nodeclaim/lifecycle/launch.go#L106) appears to needs the ID for logging at the least.

**How important is this feature to you?**

For Cluster API provider this feature is very important as without an asynchronous method for updating the provider ID we will need to use an alternative method such as a temporary sentinel value or a blocking call in the create method.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by elmiko** [comment]
cc @jonathan-innis

**Comment by jonathan-innis** [comment]
> appears to needs the ID for logging at the least

Yeah, even more critical than the logging is that this is the mechanism that Karpenter uses to determine that it's cluster state  is "synced" and it can continue to provision more nodes since it know all of the details of the nodes that it resolves. 

When you are asking for async provider id propagation, does this also involve not knowing other details about the node including what its instance type is, the number of allocatable resources, etc.? Without knowing these details, it's probably more appropriate for Karpenter to wait until it has these details for making any other decisions.

**Comment by elmiko** [comment]
> Yeah, even more critical than the logging is that this is the mechanism that Karpenter uses to determine that it's cluster state is "synced" and it can continue to provision more nodes since it know all of the details of the nodes that it resolves.

how does it use the provider id for this?

as part of my exploration, i'm thinking about the possibility of using sentinel values in the `NodeClaim.Status.ProviderID`. basically the idea would be that i could use a randomly generated sentinel value on create (eg `clusterapi://new-instance-<random numbers>`) in conjunction with an annotation on the NodeClaim designating the cluster api resource that the new Machine is owned by, then have a controller reconciling new Machine objects from cluster-api, when it sees a new Machine that is not yet claimed by karpenter, but has been created by a request from karpenter, it will then reconcile the provider id to an appropriate NodeClaim.

would changing the `NodeClaim.Status.ProviderID` after it has been created cause any issues for the core karpenter? (eg the cloudprovider.Create method is called, a fake providerID is returned, later the real providerID is updated)

> When you are asking for async provider id propagation, does this also involve not knowing other details about the node including what its instance type is, the number of allocatable resources, etc.?

i hope not, but it is a possibility. currently, i am trying to front load as much data as possible so that karpenter can have things like instance size/type during the `Create` and `GetInstanceTypes` calls. cluster api does not currently have a strong mechanism for representing this type of data though, so in the future it might be beneficial to have data propagation that is more tolerant to asynchronous behavior.

**Comment by elmiko** [comment]
would love to have more discussion on this topic to understand if there is a possibility to improve the interaction with a provider. the cluster-api provider currently uses a blocking implementation during the `Create` interface to acquire the provider id, but this feels brittle to me.

it would be nice to have a way to let karpenter core know that a node claim will have a provider id but that it is not available yet. i would imagine that the the node claim would just sit with the `registered` condition as false until the provider id is supplied, but it seems like there might be some internal map/cache that wants to use the provider id as an identifier. i'm not sure what the best solution might be, but i'd be happy to look through some code if someone could point me in the right direction =)

**Comment by elmiko** [comment]
i'm planning to do some experimentation around this topic to see what the current code tolerates. will report back here with what i find.

/remove-lifecycle stale

**Comment by elmiko** [comment]
thanks @engedaam hopefully i will have some results early in january

**Comment by elmiko** [comment]
i'm fine to let this close. i'm not sure it's a problem we need to solve in the core of karpenter.

it's worth noting that we have been using an annotation on the nodeclaim as an experiment to to address asynchronous provider ID creation, and it is working well in the Cluster API provider.
